dawn/src/dawn/CodeGen/CXXNaive-ico/CXXNaiveCodeGen.cpp

```diff
@@ -247,28 +246,56 @@ void CXXNaiveIcoCodeGen::generateStencilWrapperCtr(
-                                                   iir::FieldAccessType::StencilTemporary>()) {
-
-      if(metadata.isMultiVersionedField(accessID)) {
-        int originalAccessID = metadata.getOriginalVersionOfAccessID(accessID);
-        StencilWrapperConstructor.addInit("m_" + metadata.getNameFromAccessID(accessID) +
-                                          "(allocateFieldLike(LibTag{}, " +
-                                          metadata.getNameFromAccessID(originalAccessID) + "))");
-      } else {
-        StencilWrapperConstructor.addInit("m_" + metadata.getNameFromAccessID(accessID) +
-                                          "(allocateFieldLike(LibTag{}, " +
-                                          metadata.getNameFromAccessID(accessID) + "))");
+    if(metadata.hasAccessesOfType<iir::FieldAccessType::InterStencilTemporary,
+                                  iir::FieldAccessType::StencilTemporary>()) {
+      for(auto accessID : metadata.getAccessesOfType<iir::FieldAccessType::InterStencilTemporary,
+                                                     iir::FieldAccessType::StencilTemporary>()) {
+
+        if(metadata.isMultiVersionedField(accessID)) {
+          int originalAccessID = metadata.getOriginalVersionOfAccessID(accessID);
+          StencilWrapperConstructor.addInit("m_" + metadata.getNameFromAccessID(accessID) +
+                                            "(allocateFieldLike(LibTag{}, " +
+                                            metadata.getNameFromAccessID(originalAccessID) + "))");
+        } else {
+          auto field = stencilFields.at(accessID).field;
+          std::string allocString;
+          if(field.getFieldDimensions().isVertical()) {
+            allocString = "allocateField(LibTag{}, k_size)";
+          } else {
+            auto hdims = sir::dimension_cast<sir::UnstructuredFieldDimension const&>(
+                field.getFieldDimensions().getHorizontalFieldDimension());
+
+            auto getNumElCall = [](const sir::UnstructuredFieldDimension& hdims) -> std::string {
+              switch(hdims.getDenseLocationType()) {
+              case ast::LocationType::Cells:
+                return "numCells(LibTag{}, mesh)";
+              case ast::LocationType::Edges:
+                return "numEdges(LibTag{}, mesh)";
+              case ast::LocationType::Vertices:
+                return "numVertices(LibTag{}, mesh)";
+              default:
+                dawn_unreachable("invalid location");
+              }
+            };
+
+            if(hdims.isDense()) {
+              allocString = "allocateField(LibTag{}, " + getNumElCall(hdims) + ", k_size)";
+            } else {
+              allocString = "allocateField(LibTag{}, " + getNumElCall(hdims) + ", k_size, " +
+                            std::to_string(ICOChainSize(hdims.getNeighborChain())) + ")";
+            }
+
+            StencilWrapperConstructor.addInit("m_" + metadata.getNameFromAccessID(accessID) + "(" +
```

Comment:
Is `StencilWrapperConstructor` a variable name? it's confusing because it looks like a call to a static method given that it's capitalized.